File: src/core/masterkey.ts
import { promises as fs } from "node:fs";
import path from "node:path";
import { homedir } from "node:os";
import { writeFileSecure } from "./permissions.js";

export interface MasterKeyProvider {
  exists(): Promise<boolean>;
  read(): Promise<Buffer>;
  write(key: Buffer): Promise<void>;
  delete(): Promise<void>;
}

export class FileMasterKeyProvider implements MasterKeyProvider {
  private projectId: string;
  private basePath?: string;

  constructor(projectId: string, basePath?: string) {
    this.projectId = projectId;
    this.basePath = basePath;
  }

  private get filePath(): string {
    const base = this.basePath ?? path.join(homedir(), ".vibelock", "keys");
    return path.join(base, this.projectId, "master.key");
  }

  get keyPath(): string {
    return this.filePath;
  }

  get keyDir(): string {
    return path.dirname(this.filePath);
  }

  async exists(): Promise<boolean> {
    try {
      const stat = await fs.stat(this.filePath);
      return stat.size === 32;
    } catch {
      return false;
    }
  }

  async read(): Promise<Buffer> {
    let data: Buffer;
    try {
      data = await fs.readFile(this.filePath);
    } catch {
      throw new Error(`Master key file not found: ${this.filePath}`);
    }
    if (data.length !== 32) {
      throw new Error(
        `Invalid master key length: expected 32 bytes, got ${data.length}`
      );
    }
    return data;
  }

  async write(key: Buffer): Promise<void> {
    await writeFileSecure(this.filePath, key);
  }

  async delete(): Promise<void> {
    await fs.rm(this.filePath);
  }
}

export function getMasterKeyProvider(
  projectId?: string,
  options?: { keysPath?: string }
): MasterKeyProvider {
  const id = projectId ?? "default";
  const basePath = options?.keysPath ?? process.env.VIBELOCK_KEYS_PATH;
  return new FileMasterKeyProvider(id, basePath);
}

File: python-sdk/src/vibelock/masterkey.py
import os
from pathlib import Path
from abc import ABC, abstractmethod
from typing import Optional


class MasterKeyNotFoundError(Exception):
    pass


class MasterKeyWriteError(Exception):
    pass


class MasterKeyReadError(Exception):
    pass


class MasterKeyProvider(ABC):
    @abstractmethod
    def exists(self) -> bool: ...

    @abstractmethod
    def read(self) -> bytes: ...

    @abstractmethod
    def write(self, key: bytes) -> None: ...

    @abstractmethod
    def delete(self) -> None: ...


class FileMasterKeyProvider(MasterKeyProvider):
    def __init__(self, project_id: str = "default", base_path: Optional[Path] = None):
        self.project_id = project_id
        self.base_path = base_path or Path.home() / ".vibelock" / "keys"
        self.key_path = self.base_path / project_id / "master.key"

    def exists(self) -> bool:
        return self.key_path.exists()

    def read(self) -> bytes:
        if not self.exists():
            raise MasterKeyNotFoundError(f"Master key not found: {self.key_path}")

        try:
            return self.key_path.read_bytes()
        except Exception as e:
            raise MasterKeyReadError(f"Failed to read master key: {e}") from e

    def write(self, key: bytes) -> None:
        if len(key) != 32:
            raise ValueError("Master key must be 32 bytes")

        self.base_path.mkdir(parents=True, exist_ok=True)
        self.key_path.parent.mkdir(parents=True, exist_ok=True)

        os.chmod(self.base_path, 0o700)
        os.chmod(self.key_path.parent, 0o700)

        temp_path = self.key_path.with_suffix(".tmp")

        try:
            temp_path.write_bytes(key)

            os.chmod(temp_path, 0o600)

            os.replace(temp_path, self.key_path)

        except Exception as e:
            if temp_path.exists():
                temp_path.unlink()
            raise MasterKeyWriteError(f"Failed to write master key: {e}") from e

    def delete(self) -> None:
        if self.exists():
            try:
                self.key_path.unlink()
            except Exception as e:
                raise MasterKeyWriteError(f"Failed to delete master key: {e}") from e


def get_master_key_provider(
    project_id: str = "default",
    keys_path: str | None = None,
) -> MasterKeyProvider:
    base_path = None
    if keys_path:
        base_path = Path(keys_path)

    return FileMasterKeyProvider(project_id, base_path)
